src/R/Components/Impl/InteractiveWorkflow/Implementation/RSessionCallback.cs

```diff
@@ -62,6 +62,22 @@ internal sealed class RSessionCallback : IRSessionCallback {
             history.PlotContentProvider.LoadFile(filePath);
         }
 
+        public async Task<LocatorResult> Locator(CancellationToken ct) {
+            var historyProvider = _coreShell.ExportProvider.GetExportedValue<IPlotHistoryProvider>();
+            var history = historyProvider.GetPlotHistory(_session);
+            var tcs = new TaskCompletionSource<LocatorResult>();
+            _coreShell.DispatchOnUIThread(() => {
+                if (history.PlotContentProvider.Locator != null) {
+                    history.PlotContentProvider.Locator.StartLocatorMode(ct, tcs);
```

Comment:
It is usually a design flaw if we have to pass `TaskCompletionSource` as a parameter.
Here it would be enough to make `StartLocatorMode` async:
```{csharp}
public async Task<LocatorResult> Locator(CancellationToken ct) {
    var historyProvider = _coreShell.ExportProvider.GetExportedValue<IPlotHistoryProvider>();
    var history = historyProvider.GetPlotHistory(_session);
    return history.PlotContentProvider.Locator != null 
        ? history.PlotContentProvider.Locator.StartLocatorModeAsync(ct, tcs)
        : Task.FromResult(new LocatorResult());
}
```
and add call of `ICoreShell.SwitchToMainThreadAsync()` to the beginning of `StartLocatorModeAsync` implementation